gtsam/sfm/DsfTrackGenerator.h

```diff
@@ -0,0 +1,207 @@
+  // We create a singleton for union-find set elements from camera index of a
+  // detection and the index of that detection in that camera's keypoint list,
+  // i.e. (i,k).
+  // @param Map from (i1,i2) image pair indices to (K,2) matrix, for K
+  //        correspondence indices, from each image.
+  // @param Length-N list of keypoints, for N images/cameras.
+  std::vector<SfmTrack2d> generate_tracks_from_pairwise_matches(
+    const MatchIndicesMap& matches_dict,
+    const KeypointsList& keypoints_list) {
+    std::vector<SfmTrack2d> track_2d_list;
+
+    std::cout << "[SfmTrack2d] Starting Union-Find..." << std::endl;
+    // Generate the DSF to form tracks.
+    DSFMapIndexPair dsf;
+
+    for (const auto& kv : matches_dict) {
+      const auto pair_idxs = kv.first;
+      const auto corr_idxs = kv.second;
+
+      // Image pair is (i1,i2).
+      size_t i1 = pair_idxs.first;
+      size_t i2 =  pair_idxs.second;
+      for (size_t k = 0; k < corr_idxs.rows(); k++)
+      {
+        // Measurement indices are found in a single matrix row, as (k1,k2).
+        size_t k1 = corr_idxs(k, 0);
+        size_t k2 = corr_idxs(k, 1);
+        // Unique keys for the DSF are (i,k), representing keypoint index in an image.
+        dsf.merge(IndexPair(i1, k1), IndexPair(i2, k2));
+      }
+    }
+
+    std::cout << "[SfmTrack2d] Union-Find Complete" << std::endl;
+    const std::map<IndexPair, std::set<IndexPair> > key_sets = dsf.sets();
+
+    // Return immediately if no sets were found.
+    if (key_sets.empty()) return track_2d_list;
+
+    size_t erroneous_track_count = 0;
+    // Create a list of tracks.
+    // Each track will be represented as a list of (camera_idx, measurements).
+    for (const auto& kv : key_sets) {
+      const auto set_id = kv.first;
+      const auto index_pair_set = kv.second;
+
+      // Initialize track from measurements.
+      SfmTrack2d track_2d;
+
+      for (const auto& index_pair : index_pair_set)
+      {
+        // Camera index is represented by i, and measurement index is represented by k.
+        size_t i = index_pair.i();
+        size_t k = index_pair.j();
+        // Add measurement to this track.
+        track_2d.addMeasurement(NamedSfmMeasurement(i, keypoints_list[i].coordinates.row(k)));
+      }
+
+      // Skip erroneous track that had repeated measurements within the same image.
+      // This is an expected result from an incorrect correspondence slipping through.
+      if (track_2d.hasUniqueCameras())
+      {
+        track_2d_list.emplace_back(track_2d);
+      } else {
+        erroneous_track_count++;
+      }
+    }
+
+    double erroneous_track_pct = static_cast<float>(erroneous_track_count)
+                                 / static_cast<float>(key_sets.size()) * 100;
+    // TODO(johnwlambert): restrict decimal places to 2 decimals.
+    std::cout << "DSF Union-Find: " << erroneous_track_pct;
+    std::cout << "% of tracks discarded from multiple obs. in a single image." << std::endl;
```

Comment:
same comment, do we need this is a low-level API ?

Reply:
Thanks, added a `verbose` flag to the method that defaults to false that controls the logging verbosity.